### ISSUE TITLE
Windows spawnProcess uses invalid open mode "u"

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3567]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3567|
|Type|defect|
|Created|2008-12-05 21:43:42Z|
|Branch|https://github.com/twisted/twisted/tree/spawnprocess-file-mode-3567|

One test exercises `twisted.internet._dumbwin32proc._findShebang` apparently:

```
===============================================================================
[ERROR]: twisted.web2.test.test_cgi.CGIDirectoryTest.test_scriptsExecute

Traceback (most recent call last):
  File "C:\buildbot\twisted\winxp32-py2.6-select\Twisted\twisted\internet\defer.py", line 106, in maybeDeferred
    result = f(*args, **kw)
  File "C:\buildbot\twisted\winxp32-py2.6-select\Twisted\twisted\web2\resource.py", line 85, in renderHTTP
    return method(request)
  File "C:\buildbot\twisted\winxp32-py2.6-select\Twisted\twisted\web2\resource.py", line 128, in http_GET
    return self.render(request)
  File "C:\buildbot\twisted\winxp32-py2.6-select\Twisted\twisted\web2\twcgi.py", line 147, in render
    return runCGI(request, self.filename)
  File "C:\buildbot\twisted\winxp32-py2.6-select\Twisted\twisted\web2\twcgi.py", line 125, in runCGI
    reactor.spawnProcess(proc, filterscript, qargs, env, os.path.dirname(filename))
  File "C:\buildbot\twisted\winxp32-py2.6-select\Twisted\twisted\internet\posixbase.py", line 235, in spawnProcess
    return Process(self, processProtocol, executable, args, env, path)
  File "C:\buildbot\twisted\winxp32-py2.6-select\Twisted\twisted\internet\_dumbwin32proc.py", line 185, in __init__
    sheb = _findShebang(command)
  File "C:\buildbot\twisted\winxp32-py2.6-select\Twisted\twisted\internet\_dumbwin32proc.py", line 77, in _findShebang
    f = file(filename, 'ru')
exceptions.IOError: [Errno 22] invalid mode ('ru') or filename: 'C:\\buildbot\\twisted\\winxp32-py2.6-select\\Twisted\\_trial_temp\\twisted.web2.test.test_cgi\\CGIDirectoryTest\\test_scriptsExecute\\cmvqax\\temp\\dummy'
-------------------------------------------------------------------------------
```

It fails on Python 2.6 on Windows, which apparently has stricter mode checking than previous versions of Python on Windows.

It seems this will prevent #! interpreters from being supported on Windows on Python 2.6, until fixed.

Attachments:

* [twisted_windows_cgi_file_mode.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f06/f06f30fa6356cae7511a4cfb1a93b5dfe0070800/2a924decd7a124fded0fa0f97dfb05adf6b4af8e.patch) (405 bytes) - added by azverkan on 2009-05-23 02:31:14Z - 
* [shebang.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f06/f06f30fa6356cae7511a4cfb1a93b5dfe0070800/99b96603c2bdcb0f4da2e23b835a6d86454f110f.patch) (1547 bytes) - added by abbeyj on 2009-07-28 04:43:49Z - 
* [shebang2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f06/f06f30fa6356cae7511a4cfb1a93b5dfe0070800/6b89933dde6aa3dce74b83ab63be29fed6830cad.patch) (2448 bytes) - added by abbeyj on 2009-07-28 20:20:16Z - Move test to twisted.internet.test.test_process

<details><summary>Searchable metadata</summary>

```
trac-id__3567 3567
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__Python_2_6 Python-2.6
branch__branches_spawnprocess_file_mode_3567 branches/spawnprocess-file-mode-3567
branch_author__abbeyj__exarkun abbeyj, exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1228513422000000 1228513422000000
changetime__1254181600000000 1254181600000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

I think that's supposed to be "rU", i.e. "universal newline mode".

**Comment by twisted-trac** [comment]
|<img alt="azverkan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|azverkan commented|
|-|-|

Add review keyword to go with the patch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to azverkan|
|-|-|

Thanks for the patch.  Though there's already a failing unit test, it's only a somewhat accidental test.  There needs to be a new unit test directly for this code.

**Comment by twisted-trac** [comment]
|<img alt="abbeyj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|abbeyj commented|
|-|-|

Here's an attempt at a test.  I'm not sure if this is the right place to put it but it was convenient.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Hi abbeyj,

Thanks for the patch.  This looks like a pretty decent test for the functionality.  There are two points about it I'd like to mention, though.

First, the "place" - there is indeed somewhere which would be a bit better for it, `twisted.internet.test.test_process`.  Aside from organization, this would be beneficial because `twisted.internet.test` has infrastructure which allows tests to be run against all reactors supported on the host platform in a single test run.  Tests from places like `twisted.test.test_iutils` predate this infrastructure and to exercise multiple reactors, trial must be run multiple times, each time with a different argument for `--reactor`.

Second, it would be cool to unit test this basic reactor functionality by invoking it directly - ie, not via a higher level helper function.  This produces a better unit test (since less code is involved) and makes the code more closely reflect the intent of the test.  Of course, the drawback is that writing this test without `getProcessOutputAndValue` means it's quite a bit more code.  In this case, the drawback may outweigh the benefits, so leaving it as it is, using `getProcessOutputAndValue`, may be the best thing.

Do you want to have a crack at updating the patch to address the first point, though?

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|ivank commented|
|-|-|

adding review keyword

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|ivank removed owner|
|-|-|

assigning to no one

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to abbeyj|
|-|-|

Hi abbeyj, thanks for your work here.  Sorry to give you a kind of lame review after a month has gone by, but I'm a little confused as to what's actually up for review.  It looks like `shebang2.patch` is missing the implementation of the functionality, and includes only a test... which is different from the test in `shebang.patch` and in a different place.

Can you attach a single patch which includes all of your proposed changes at once?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [27308]) Branching to 'spawnprocess-file-mode-3567'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [27309]) Apply part of shebang.patch and all of shebang2.patch

refs [#3567](https://github.com/twisted/twisted/issues/3567)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [27310]) Minor tweaks to the test docstring and the reactor management

refs [#3567](https://github.com/twisted/twisted/issues/3567)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

I took the patches and applied the parts of them which seemed to make sense to the branch mentioned above.  The code looked pretty good to me, I just made a couple minor tweaks.  Someone else should give it a quick glance.  Here are the [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/spawnprocess-file-mode-3567)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_13"></a>|@jesstess commented|
|-|-|

twisted/internet/test/test_process.py needs a copyright bump.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [27343]) Update copyright date

refs [#3567](https://github.com/twisted/twisted/issues/3567)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set owner to @exarkun|
|-|-|

```
+        Spawning a process with an exexutable which is a script starting
```

Typo: executable.

Looks good to me, +1!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to @therve|
|-|-|

Thanks.  Can you fix the typo and merge for me?

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_17"></a>|@radix set status to `closed`|
|-|-|

(In [27356]) Merge spawnprocess-file-mode-3567

Author: abbeyj, exarkun
Reviewer: therve, jesstess
Fixes: [#3567](https://github.com/twisted/twisted/issues/3567)

Executing a "script" on Windows (i.e., a file which starts with a #!
line) in Python 2.6 now works after a change to read the script file with the
real mode "rU" instead of the non-mode "ru". In earlier versions of Python on
Windows, unknown mode letters were ignored.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
